### ISSUE TITLE
Train 3D nnU-Net

### ISSUE THREAD

**Issue by plbenveniste** [ivadomed/canproco#29]
- [x] Explore and understand the CanProCo dataset then select the data to use for training
- [x] Data-preprocessing for nnU-Net format
- [x] Code orientation correction script
- [x] Train a 3D nnU-Net
- [x] Test on untouched data (maybe the [data from the second-time point](https://github.com/ivadomed/canproco/issues/30))
- [x] Analysis of the results

Use code from [Nilser's work](https://github.com/brainhack-school2023/laines_project)

**Comment by plbenveniste** [comment]
Data pre-processing done: (445 training volumes, 1 test volume)
Only `sub-mon119_ses-M0_PSIR.nii.gz` is not annotated in CanProCo.

**Comment by valosekj** [comment]
> Only `sub-mon119_ses-M0_PSIR.nii.gz` is not annotated in CanProCo.

Good catch! I do not know why `sub-mon119_ses-M0_PSIR.nii.gz` does not have a lesion mask. I checked the image; it looks okay, and the lesion is visible. I guess we can segment the lesion and upload it to git-annex.

**Comment by plbenveniste** [comment]
During nnU-Net preprocessing, I got the following warning message "Direction mismatch between segmentation and corresponding images." for the following image (99, 101, 102, 104, 105, 094, 107, 108, 097, 127, 155, 141, 128, 111, 142, 115, 112, 143, 116, 130, 144, 117, 159, 145, 118, 132, 146, 119, 147, 120, 134, 162, 148, 121, 135, 122, 177, 123, 124, 152, 139, 125, 180, 153, 126 : cf [conversion_dict.json.zip(https://github.com/ivadomed/canproco/files/11839812/conversion_dict.json.zip)). 

After observing some of these images on FSLeyes, i got the warning: "Displaying images with different orientations/fields of view". 
Indeed, the orientations are not the same and therefore the segmentations are outside the image volume (see white mark bottom right of the image below): 
![image](https://github.com/ivadomed/canproco/assets/67429280/ae99bea2-23ec-42b6-91c5-e1642e4be543)

**Comment by valosekj** [comment]
Thanks for reporting this! It might be related to this issue: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4025. 
Especially see this (hidden) comment: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4025#issuecomment-1560173129

**Comment by plbenveniste** [comment]
The above image is : `sub-edm018_ses-M0_PSIR.nii.gz. The conversion from number to subjectID can be done using the `conversion_dict.json` above.

**Comment by plbenveniste** [comment]
Updating FSLeyes to 1.7.0 corrected the visualisation problem. However, since nnUNet is sensitive to that, we will correct the orientation of the above-listed images.

**Comment by plbenveniste** [comment]
2 images still remain problematic : 
![image](https://github.com/ivadomed/canproco/assets/67429280/44d02390-498a-47e4-b9fe-3d96464374f7)

Because of the very small difference in orientation, I will ignore it for the training

**Comment by plbenveniste** [comment]
Results from training on `sub-mon119_ses-M0_PSIR.nii.gz` (which isn't labelled) before correcting the file orientations:
![canproco_mon119](https://github.com/ivadomed/canproco/assets/67429280/c75c4bbf-b57d-4ac7-8800-ff0f5a85d4d2)

(Training currently in progress with corrected files)

**Comment by plbenveniste** [comment]
Resuts from fold 0 and 1 (max Dice =0.47): not great -> currently training fold 2,3 and 4. 
Example on image 002 (`sub-cal072_ses-M0_STIR.nii.gz`)
![gif_canproco_002](https://github.com/ivadomed/canproco/assets/67429280/9ed17eba-e063-4de8-b966-c913d444c250)
In this case, the models "over-segment". 

Image 210 (`sub-mon129_ses-M0_PSIR.nii.gz`): Model "under-segment"
Image 385 (`sub-van144_ses-M0_PSIR.nii.gz`): model misses the only lesion (but it is small)
Image 398 (`sub-van160_ses-M0_PSIR.nii.gz`): model misses some of the lesions

It seems that overall the modle favours precision over recall. How can we modify a threshold for it to be less selective ? 
Looking into this [issue](https://github.com/MIC-DKFZ/nnUNet/discussions/922).

**Comment by jcohenadad** [comment]
@plbenveniste are you training a model that combines PSIR and STIR? not the _very_ different contrast between the two (CSF white vs. dark). If you did, I find these results quite promising (Dice on MS lesion seg task is expected to be around 0.5, given the very high inter-rater variability), and I would encourage you to maybe consider aggregating with other contrasts (eg MP2RAGE, T2w)

**Comment by plbenveniste** [comment]
The nnUNet model is currently trained on both PSIR and STIR data. After full training on all 5 folds, we will try training with other contrasts as well.

**Comment by valosekj** [comment]
FYI - I moved the discussion about the missing lesion mask for `sub-mon119_ses-M0_PSIR.nii.gz` into https://github.com/ivadomed/canproco/issues/34.

**Comment by plbenveniste** [comment]
Here are the results of the training on the 5 folds: 
<img width="946" alt="image" src="https://github.com/ivadomed/canproco/assets/67429280/f06edab3-e740-4399-9f73-be6762e7c2bd">
The same had happened during the previous training on only one fold (fold 5) : Best Dice : 0.5077 : 
<img width="713" alt="image" src="https://github.com/ivadomed/canproco/assets/67429280/60bc390d-f17b-45eb-8a3a-41fcdffe02ab">

We can see strange patterns for the Dice Loss which suddenly drops at some point during training. 
Currently looking into this issue.

**Comment by plbenveniste** [comment]
One hypothesis which could explain the previous training curve is that 80% of the dataset consists of PSIR contrast and only 20% consists of TSIR. Furthermore, the two previous contrasts are pretty opposite in this case as, for PSIR lesions will be hypointense while for TSIR lesions will be hyperintense. 

Therefore, I want to try using only PSIR images to train a 3D nnU-Net model to see how it performs. 

Tasks:

- [ ] create new training and testing dataset
- [ ] preprocess to 3d nnU-Net format
- [ ] train 3d nnU-Net on 1 fold
- [ ] look at the training curve
- [ ] look at the results on validation dataset
- [ ] perform inference on M12 dataset

**Comment by plbenveniste** [comment]
Here are the training curve for fold 1 and fold 2: 

For fold 1: 
<img width="527" alt="image" src="https://github.com/ivadomed/canproco/assets/67429280/762bb258-f698-475e-8a6d-1513fdc73e73">

For fold 2: 
<img width="525" alt="image" src="https://github.com/ivadomed/canproco/assets/67429280/9c788ab9-341c-4aab-8794-46739ab4f1b0">

It seems that the homogeneity of the dataset wasn't causing the drop in training curve. One might say that it could be linked to the heterogeneity between sites but it wouldn't make sense for us to train a specific model for each site. 

Solution : 
- Now moving away from nnUNet and UNet based model as they seem not to be performing very well for this task. 
- Currently looking into diffusion models

**Comment by plbenveniste** [comment]
Closing this issue as the model was integrated into SCT.
